metricflow/naming/dunder_scheme.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+import re
+from typing import Optional, Sequence, Tuple
+
+from dbt_semantic_interfaces.naming.keywords import DUNDER
+from dbt_semantic_interfaces.references import EntityReference
+from dbt_semantic_interfaces.type_enums import TimeGranularity
+from dbt_semantic_interfaces.type_enums.date_part import DatePart
+from typing_extensions import override
+
+from metricflow.naming.naming_scheme import QueryItemNamingScheme
+from metricflow.specs.patterns.entity_link_pattern import (
+    EntityLinkPattern,
+    EntityLinkPatternParameterSet,
+    ParameterSetField,
+)
+from metricflow.specs.specs import (
+    LinkableInstanceSpec,
+    LinkableSpecSet,
+    LinkableSpecSetTransform,
+)
+
+
+class DunderNamingScheme(QueryItemNamingScheme):
+    """A naming scheme using the dundered name syntax.
+
+    TODO: Consolidate with StructuredLinkableSpecName / DunderedNameFormatter.
+    """
+
+    _INPUT_REGEX = re.compile(r"\A[a-z]([a-z0-9_])*[a-z0-9]\Z")
+
+    @staticmethod
+    def date_part_suffix(date_part: DatePart) -> str:
+        """Suffix used for names with a date_part."""
+        return f"extract_{date_part.value}"
+
+    @override
+    def input_str(self, instance_spec: LinkableInstanceSpec) -> Optional[str]:
+        spec_set = LinkableSpecSet.from_specs((instance_spec,))
+
+        for time_dimension_spec in spec_set.time_dimension_specs:
+            # From existing comment in StructuredLinkableSpecName:
+            #
+            # Dunder syntax not supported for querying date_part
+            #
+            if time_dimension_spec.date_part is not None:
+                return None
+        names = _DunderNameTransform().transform(LinkableSpecSet.from_specs((instance_spec,)))
+        if len(names) != 1:
+            raise RuntimeError(f"Did not get 1 name for {instance_spec}. Got {names}")
+
+        return names[0]
+
+    @override
+    def spec_pattern(self, input_str: str) -> EntityLinkPattern:
+        if not self.input_str_follows_scheme(input_str):
+            raise RuntimeError(f"{repr(input_str)} does not follow this scheme.")
```

Comment:
What exception gets thrown in this case today? I assume this is almost always going to be a query input error, which means we may want to throw an exception we can handle and re-cast for things like alert management.

If we don't have a custom exception we can go to here I'd tag this as ValueError since it's an invalid value. RuntimeError is really broad. Also, we use ValueError below.